src/lxc/conf.c

```diff
@@ -757,79 +788,80 @@ static int lxc_mount_auto_mounts(struct lxc_conf *conf, int flags, struct lxc_ha
 	}
 
 	if (flags & LXC_AUTO_SHMOUNTS_MASK) {
-		char *new_mount;
-		size_t len_mount;
+		size_t len_cont;
+		char *full_cont_path;
+		struct stat sb;
 		int ret = -1;
 
+		/* Construct the shmount path under the container root */
 		/* +1 for slash */
-		size_t len_cont = strlen(conf->rootfs.mount) + 1 + strlen(conf->lxc_shmount.path_cont);
+		len_cont = strlen(conf->rootfs.mount) + 1 + strlen(conf->lxc_shmount.path_cont);
 		/* +1 for the terminating '\0' */
-		char *full_cont_path = malloc(len_cont + 1);
+		full_cont_path = malloc(len_cont + 1);
 		if(!full_cont_path) {
 			SYSERROR("Not enough memory");
 			return -ENOMEM;
 		}
-
 		ret = snprintf(full_cont_path, len_cont + 1, "%s/%s", conf->rootfs.mount, conf->lxc_shmount.path_cont);
 		if (ret < 0 || ret >= len_cont + 1) {
 			SYSERROR("Failed to create filename");
-			goto out_free_path;
+			free(full_cont_path);
+			return -1;
 		}
 
+		/* Check if shmount point is already set up */
 		if(is_shared_mountpoint(conf->lxc_shmount.path_host)
-				&& is_shared_mountpoint(full_cont_path)) {
-			/* Already set up */
-			ret = 0;
-			goto out_free_path;
+				&& !stat(full_cont_path, &sb) && S_ISDIR(sb.st_mode)) {
+			/* The host path is a shared mounpoint
+			 * (TODO: not checked for MS_REC),
+			 * the container path exists --
+			 * so we only need to add it to mount_list
+			 * for the following setup */
+			goto finalize;
 		}
 
-		/* Borrowed from mount_entry_create_dir_file() */
+		/* Borrowed from mount_entry_create_dir_file()
+		 *
+		 * Create host and cont mount paths
+		 */
```

Comment:
You can remove this comment. It's ok to copy and paste stuff from within our library without comments. :)